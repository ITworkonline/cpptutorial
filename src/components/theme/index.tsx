const baseTheme = {
  fonts: {
    mono: '"SF Mono", "Roboto Mono", Menlo, monospace',
  },
};

const palette = {
  white: "#ffffff",
  light: "#E2E4E9",
  grey1: "#C4C9D4",
  grey2: "#A7AEBE",
  grey3: "#8A93A8",
  grey4: "#6C7893",
  grey5: "#576075",
  grey6: "#414858",
  dark: "#2B303B",
  black: "#0B0C0F",
  blue1: "#0B1A28",
  blue2: "#08365E",
  blue3: "#224F77",
  blue4: "#2D6A9F",
  blue5: "#3884C7",
  blue6: "#609DD2",
  blue7: "#88B5DD",
  blue8: "#AFCEE9",
  blue9: "#D7E6F4",
};

const lightTheme = {
  ...baseTheme,
  colors: {
    background: palette.white,
    headerBg: palette.light,
    icon: palette.blue5,
    iconHover: palette.blue6,
    buttonBg: "#ddd",
    buttonBg2: "#ccc",
    heading: "#000",
    primaryText: "#3B454E",
    preFormattedText: "rgb(245, 247, 249)",
    link: palette.blue5,
    highlight: palette.blue5,
    sidebarHighlight: palette.blue5,
    search: palette.blue5,
    searchPlaceholder: palette.dark,
  },
};

export type Theme = typeof lightTheme;

const darkTheme: Theme = {
  ...baseTheme,
  colors: {
    background: palette.dark,
    headerBg: palette.black,
    icon: palette.blue5,
    iconHover: palette.blue6,
    buttonBg: "#0f4c75",
    buttonBg2: "#34495e",
    heading: "#fff",
    primaryText: "#fff",
    preFormattedText: "#000",
    link: palette.blue5,
    highlight: palette.blue5,
    sidebarHighlight: palette.blue3,
    search: palette.blue5,
    searchPlaceholder: palette.light,
  },
};

export { lightTheme, darkTheme };
