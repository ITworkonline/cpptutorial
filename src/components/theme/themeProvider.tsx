import * as React from "react";
import { ThemeProvider as EmotionThemeProvider } from "emotion-theming";

import { lightTheme, darkTheme } from "./index";
import Header from "../header";

interface ThemeProviderProps {
  children: React.ReactNode;
}

interface ThemeProviderState {
  isDarkThemeActive: boolean;
}

class ThemeProvider extends React.Component<ThemeProviderProps, ThemeProviderState> {
  state = {
    isDarkThemeActive: false,
  };

  componentDidMount() {
    this.retrieveActiveTheme();
  }

  retrieveActiveTheme = () => {
    const isDarkThemeActiveJson = window.localStorage.getItem("isDarkThemeActive");

    const isDarkThemeActive = isDarkThemeActiveJson ? JSON.parse(isDarkThemeActiveJson) : false;

    this.setState({ isDarkThemeActive });
  };

  toggleActiveTheme = () => {
    this.setState(prevState => ({ isDarkThemeActive: !prevState.isDarkThemeActive }));

    window.localStorage.setItem("isDarkThemeActive", JSON.stringify(!this.state.isDarkThemeActive));
  };

  render() {
    const { children } = this.props;

    const { isDarkThemeActive } = this.state;

    const currentActiveTheme = isDarkThemeActive ? darkTheme : lightTheme;

    return (
      <div>
        <EmotionThemeProvider theme={currentActiveTheme}>
          <Header
            isDarkThemeActive={isDarkThemeActive}
            toggleActiveTheme={this.toggleActiveTheme}
          />
          {children}
        </EmotionThemeProvider>
      </div>
    );
  }
}

export default ThemeProvider;
