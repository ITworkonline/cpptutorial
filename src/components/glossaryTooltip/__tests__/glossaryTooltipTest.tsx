import React from "react";
import { render, cleanup } from "@testing-library/react";

import "@testing-library/jest-dom";
import renderer from "react-test-renderer";

import { ThemeProvider } from "emotion-theming";
import { lightTheme } from "../../theme";

import GlossaryTooltip from "..";

jest.mock("gatsby-plugin-mdx", () => {
  return {
    MDXRenderer: ({ children, ...props }) => <div {...props}>{children}</div>,
  };
});

beforeEach(cleanup);

const title = "std::puts";
const tooltipType = "C-style IO Library";
const body = `return () => React.createElement('div')`;
const cppreference = "cpp/io/c/puts";

const props = { title, type: tooltipType, cppreference };

it("Renders correctly without cppreference", () => {
  const { getByTestId, queryByTestId } = render(
    <ThemeProvider theme={lightTheme}>
      <GlossaryTooltip title={title} type={tooltipType}>
        {body}
      </GlossaryTooltip>
    </ThemeProvider>
  );

  const tooltipElem = getByTestId("glossary-tooltip");
  const headingElem = getByTestId("glossary-tooltip-heading");
  const titleElem = getByTestId("glossary-tooltip-title");
  const tooltipTypeElem = getByTestId("glossary-tooltip-type");
  const bodyElem = getByTestId("glossary-tooltip-body");
  const cppreferenceElem = queryByTestId("glossary-tooltip-cppreference");

  expect(tooltipElem).toContainElement(headingElem);
  expect(headingElem).toContainElement(titleElem);
  expect(titleElem).toHaveTextContent(title);
  expect(headingElem).toContainElement(tooltipTypeElem);
  expect(tooltipTypeElem).toHaveTextContent(tooltipType);
  expect(tooltipElem).toContainElement(bodyElem);

  expect(cppreferenceElem).toBeNull();
});

it("Renders correctly with cppreference", () => {
  const { getByTestId } = render(
    <ThemeProvider theme={lightTheme}>
      <GlossaryTooltip {...props}>{body}</GlossaryTooltip>
    </ThemeProvider>
  );

  const tooltipElem = getByTestId("glossary-tooltip");
  const headingElem = getByTestId("glossary-tooltip-heading");
  const titleElem = getByTestId("glossary-tooltip-title");
  const tooltipTypeElem = getByTestId("glossary-tooltip-type");
  const bodyElem = getByTestId("glossary-tooltip-body");
  const cppreferenceElem = getByTestId("glossary-tooltip-cppreference");

  expect(tooltipElem).toContainElement(headingElem);
  expect(headingElem).toContainElement(titleElem);
  expect(titleElem).toHaveTextContent(title);
  expect(headingElem).toContainElement(tooltipTypeElem);
  expect(tooltipTypeElem).toHaveTextContent(tooltipType);
  expect(tooltipElem).toContainElement(bodyElem);

  expect(cppreferenceElem).toHaveTextContent("See cppreference for detailed documentation");
});

it("Passes snapshot test", () => {
  const tree = renderer
    .create(
      <ThemeProvider theme={lightTheme}>
        <GlossaryTooltip {...props}>{body}</GlossaryTooltip>
      </ThemeProvider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
