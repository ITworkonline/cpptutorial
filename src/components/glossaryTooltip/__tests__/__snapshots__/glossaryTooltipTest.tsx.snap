// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Passes snapshot test 1`] = `
.emotion-0 {
  background-color: #ffffff;
  color: #3B454E;
  border: 5px dashed red;
  border-radius: 6px;
  opacity: 0.95;
  white-space: normal;
  word-wrap: break-word;
  position: absolute;
  z-index: 1;
  padding: 10px 20px;
  width: 100%;
  left: 0;
}

.emotion-0 p {
  margin: 8px 0px;
}

.emotion-0 .title h2 {
  margin-bottom: 5px;
}

.emotion-0 .title span {
  color: #3884C7;
}

@media (min-width:768px) {
  .emotion-0 {
    width: auto;
    left: auto;
    position: absolute;
    max-width: 600px;
  }

  .emotion-0 .title h2 {
    display: inline-block;
  }

  .emotion-0 .title span {
    float: right;
  }
}

.emotion-0:hover {
  visibility: visible;
}

<article
  className="emotion-0 emotion-1"
  data-testid="glossary-tooltip"
>
  <div
    data-testid="glossary-tooltip-heading"
  >
    <h2
      data-testid="glossary-tooltip-title"
    >
      std::puts
    </h2>
    <span
      data-testid="glossary-tooltip-type"
    >
      C-style IO Library
    </span>
  </div>
  <div
    data-testid="glossary-tooltip-body"
  >
    return () =&gt; React.createElement('div')
  </div>
  <p
    data-testid="glossary-tooltip-cppreference"
  >
    See 
    <a
      href="https://en.cppreference.com/w/cpp/io/c/puts"
    >
      cppreference
    </a>
     for detailed documentation
  </p>
</article>
`;
